src/Npgsql/NpgsqlReadBuffer.cs

```diff
@@ -96,60 +96,60 @@ internal void Ensure(int count)
 
         internal Task Ensure(int count, bool async, bool dontBreakOnTimeouts)
         {
-            return count <= ReadBytesLeft ? Task.CompletedTask : EnsureLong();
+            return count <= ReadBytesLeft ? Task.CompletedTask : EnsureLong(count, async, dontBreakOnTimeouts);
+        }
+
+        async Task EnsureLong(int count, bool async, bool dontBreakOnTimeouts)
```

Comment:
Better to make it `static` local method, so it won't allocate the state, but still will be accessible only by `Ensure`.

Reply:
`static` won't work since we access `this.`. 

I can make it local and introduce some prefix for arg names to avoid name-collision warning. WDYT?

Reply:
You can pass `this` as `NpgsqlReadBuffer self`, so it will be the same, but don't think it's very important. Mostly a styling issue (:

Let's hear what @roji thinks on this.

Reply:
Made it local, double-checked that fix is still valid (no excessive allocations).
Will wait for your decision on `this` vs instance argument.

What do you think, should i rebase fix on `main` branch, or it still makes sense for 4.x?

Reply:
Thanks for this improvement @Belorus! There are several other cases like this in the codebase, #3315 tracks this kind of work.

I used to be against static local methods because of the `this`/`self` thing, but now I think I agree with @YohDeadfall. It's weird stylistically, but doing this it ensures that we don't accidentally introduce a new capture in the capture and start allocating again.

However, I'd rather we just did this against main - the release is just days away, and I'd rather we left 4.x alone except for actual bug fixes...

Reply:
@roji means that we do not accept merges to `stable`, only to `main` and in some cases to `hotfix` branches (for example, a bug present only in a specific version).

Reply:
Thanks @roji

Will rebase on main, and try converting to C# 9 static lambda (haven't done this before :))

Reply:
You don't need a C# 9 static lambda - just a static local method (this exists in C# 8)

Reply:
Static lambdas is `static () => { }` (previously there was no way to mark it `static` to prevent captures), but that's a `static` local method available since C# 7.

Reply:
Created new PR #3336 .
Made local function static.